src/transformStateWithClones.js

```diff
@@ -7,7 +7,42 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let clone = Object.assign({}, state);
+
+  for (const item of actions) {
+    if (item.type === 'addProperties') {
+      Object.assign(clone, item.extraData);
+
+      const newData = Object.assign({}, clone);
+
+      history.push(newData);
+    }
+
+    if (item.type === 'removeProperties') {
+      const newData = Object.assign({}, clone);
+
+      for (const key in item.keysToRemove) {
+        const keyToRemove = item.keysToRemove[key];
+
+        delete newData[keyToRemove];
+      }
+      history.push(newData);
+      clone = Object.assign({}, newData);
+    }
+
+    if (item.type === 'clear') {
+      const newData = Object.assign({}, clone);
+
+      for (const key in clone) {
+        delete newData[key];
+      }
+      history.push(newData);
+      clone = Object.assign({}, newData);
+    }
```

Comment:
[CODE STYLE]: Use switch statement if you have limited amount of conditions. In this case, you have a predefined set of action types ('addProperties', 'removeProperties', 'clear'), so a switch statement would be more appropriate than multiple if statements.